File: spring-cloud-starter/src/main/java/com/jmsoftware/maf/springcloudstarter/minio/MinioHelper.java
package com.jmsoftware.maf.springcloudstarter.minio;

import cn.hutool.core.util.ObjectUtil;
import io.minio.*;
import io.minio.http.Method;
import io.minio.messages.Bucket;
import io.minio.messages.DeleteObject;
import io.minio.messages.Item;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Component;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.multipart.MultipartFile;

import javax.validation.constraints.*;
import java.io.InputStream;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * <h1>MinioHelper</h1>
 * <p>
 * Change description here.
 *
 * @author Johnny Miller (锺俊), email: johnnysviva@outlook.com, date: 6/7/21 9:52 PM
 **/
@Slf4j
@Validated
@Component
@RequiredArgsConstructor
@SuppressWarnings("unused")
public class MinioHelper {
    /**
     * 7 days
     */
    private static final int DEFAULT_EXPIRY_TIME = 7 * 24 * 3600;
    private final MinioClient minioClient;

    @SneakyThrows
    public boolean bucketExists(@NotBlank String bucket) {
        return this.minioClient.bucketExists(BucketExistsArgs.builder().bucket(bucket).build());
    }

    @SneakyThrows
    @SuppressWarnings("UnusedReturnValue")
    public boolean makeBucket(@NotBlank String bucket) {
        if (this.bucketExists(bucket)) {
            log.warn("The bucket named '{}' exists", bucket);
            return false;
        }
        this.minioClient.makeBucket(MakeBucketArgs.builder().bucket(bucket).build());
        return true;
    }

    @SneakyThrows
    public ObjectWriteResponse put(@NotBlank String bucket, @NotBlank String object,
                                   @NotNull MultipartFile multipartFile) {
        return this.minioClient.putObject(
                PutObjectArgs.builder()
                        .bucket(bucket)
                        .object(object)
                        .stream(multipartFile.getInputStream(), multipartFile.getSize(), -1)
                        .contentType(multipartFile.getContentType())
                        .build());
    }

    @SneakyThrows
    public ObjectWriteResponse putObject(@NotBlank String bucket, @NotBlank String object,
                                         @NotNull InputStream inputStream, @NotBlank String contentType) {
        return this.minioClient.putObject(
                PutObjectArgs
                        .builder()
                        .bucket(bucket)
                        .object(object)
                        .stream(inputStream, inputStream.available(), -1)
                        .contentType(contentType)
                        .build());
    }

    @SneakyThrows
    public List<String> listBucketNames() {
        return this.listBuckets().stream().map(Bucket::name).collect(Collectors.toList());
    }

    @SneakyThrows
    public List<Bucket> listBuckets() {
        return this.minioClient.listBuckets();
    }

    @SneakyThrows
    public boolean removeBucket(@NotBlank String bucket) {
        if (!this.bucketExists(bucket)) {
            return false;
        }
        val myObjects = this.listObjects(bucket);
        for (val result : myObjects) {
            val item = result.get();
            // If item has files, fail to remove
            if (item.size() > 0) {
                return false;
            }
        }
        // If bucket is empty, then it can be removed
        this.minioClient.removeBucket(RemoveBucketArgs.builder().bucket(bucket).build());
        return !this.bucketExists(bucket);
    }

    @SneakyThrows
    public List<String> listObjectNames(@NotBlank String bucket) {
        val listObjectNames = new LinkedList<String>();
        if (!this.bucketExists(bucket)) {
            return listObjectNames;
        }
        val myObjects = this.listObjects(bucket);
        for (val result : myObjects) {
            val item = result.get();
            listObjectNames.add(item.objectName());
        }
        return listObjectNames;
    }

    @SneakyThrows
    public Iterable<Result<Item>> listObjects(@NotBlank String bucket) {
        if (!this.bucketExists(bucket)) {
            return null;
        }
        return this.minioClient.listObjects(ListObjectsArgs.builder().bucket(bucket).build());
    }

    @SneakyThrows
    public GetObjectResponse getObject(@NotBlank String bucket, @NotBlank String object) {
        if (!this.bucketExists(bucket)) {
            return null;
        }
        val statObjectResponse = this.statObject(bucket, object);
        if (ObjectUtil.isNull(statObjectResponse) || statObjectResponse.size() == 0) {
            return null;
        }
        return this.minioClient.getObject(GetObjectArgs.builder().bucket(bucket).object(object).build());
    }

    @SneakyThrows
    public GetObjectResponse getObject(@NotBlank String bucket, @NotBlank String object, @Min(0L) long offset,
                                       @NotNull @Min(0L) Long length) {
        if (!this.bucketExists(bucket)) {
            return null;
        }
        val statObjectResponse = this.statObject(bucket, object);
        if (ObjectUtil.isNull(statObjectResponse) || statObjectResponse.size() == 0) {
            return null;
        }
        return this.minioClient.getObject(
                GetObjectArgs.builder().bucket(bucket).object(object).offset(offset).length(length).build());
    }

    @SneakyThrows
    public boolean removeObjects(String bucket, String object) {
        if (!this.bucketExists(bucket)) {
            return false;
        }
        this.minioClient.removeObject(RemoveObjectArgs.builder().bucket(bucket).object(object).build());
        return true;
    }

    @SneakyThrows
    public List<String> removeObjects(String bucket, List<String> objects) {
        val deleteErrorNameList = new LinkedList<String>();
        if (this.bucketExists(bucket)) {
            val deleteObjectList = objects.stream().map(DeleteObject::new).collect(Collectors.toList());
            val results = this.minioClient.removeObjects(
                    RemoveObjectsArgs.builder().bucket(bucket).objects(deleteObjectList).build());
            for (val result : results) {
                val error = result.get();
                deleteErrorNameList.add(error.objectName());
            }
        }
        return deleteErrorNameList;
    }

    @SneakyThrows
    public String getPresignedObjectUrl(@NotBlank String bucket, @NotBlank String object,
                                        @NotNull Method method,
                                        @NotNull @Min(1L) @Max(DEFAULT_EXPIRY_TIME) Integer expiration) {
        if (!this.bucketExists(bucket)) {
            return null;
        }
        return this.minioClient.getPresignedObjectUrl(
                GetPresignedObjectUrlArgs
                        .builder()
                        .bucket(bucket)
                        .object(object)
                        .method(method)
                        .expiry(expiration)
                        .build());
    }

    @SneakyThrows
    public StatObjectResponse statObject(@NotBlank String bucket, @NotBlank String object) {
        if (!this.bucketExists(bucket)) {
            return null;
        }
        return this.minioClient.statObject(StatObjectArgs.builder().bucket(bucket).object(object).build());
    }

    @SneakyThrows
    public String getPresignedObjectUrl(@NotBlank String bucket, @NotBlank String object) {
        if (!this.bucketExists(bucket)) {
            return null;
        }
        return this.minioClient.getPresignedObjectUrl(
                GetPresignedObjectUrlArgs.builder().bucket(bucket).object(object).build());
    }

    @SneakyThrows
    public ObjectWriteResponse composeObject(@NotBlank String bucket, @NotBlank String object,
                                             @NotEmpty List<ComposeSource> sources,
                                             @Nullable Map<String, String> headers) {
        if (!this.bucketExists(bucket)) {
            return null;
        }
        return this.minioClient.composeObject(
                ComposeObjectArgs.builder().bucket(bucket).object(object).sources(sources).headers(headers).build());
    }
}

File: spring-cloud-starter/src/main/java/com/jmsoftware/maf/springcloudstarter/minio/MinioHelper.kt
package com.jmsoftware.maf.springcloudstarter.minio

import cn.hutool.core.collection.CollUtil
import com.google.common.collect.Lists
import com.jmsoftware.maf.common.util.logger
import io.minio.*
import io.minio.http.Method
import io.minio.messages.Bucket
import io.minio.messages.DeleteObject
import io.minio.messages.Item
import org.springframework.validation.annotation.Validated
import org.springframework.web.multipart.MultipartFile
import java.io.InputStream
import javax.validation.constraints.*

/**
 * # MinioHelper
 *
 * Change description here.
 *
 * @author Johnny Miller (锺俊), e-mail: johnnysviva@outlook.com, date: 4/18/22 9:20 PM
 */
@Suppress("unused")
@Validated
class MinioHelper(
    private val minioClient: MinioClient
) {
    companion object {
        /**
         * 7 days
         */
        private const val DEFAULT_EXPIRY_TIME = 7 * 24 * 3600
        private val log = logger()
    }

    fun bucketExists(bucket: @NotBlank String?): Boolean {
        return minioClient.bucketExists(BucketExistsArgs.builder().bucket(bucket).build())
    }

    fun makeBucket(bucket: @NotBlank String?): Boolean {
        if (bucketExists(bucket)) {
            MinioHelper.log.warn("The bucket named `$bucket` exists")
            return false
        }
        minioClient.makeBucket(MakeBucketArgs.builder().bucket(bucket).build())
        return true
    }

    fun put(
        bucket: @NotBlank String,
        `object`: @NotBlank String,
        multipartFile: @NotNull MultipartFile
    ): ObjectWriteResponse {
        return minioClient.putObject(
            PutObjectArgs.builder()
                .bucket(bucket)
                .`object`(`object`)
                .stream(multipartFile.inputStream, multipartFile.size, -1)
                .contentType(multipartFile.contentType)
                .build()
        )
    }

    fun putObject(
        bucket: @NotBlank String,
        `object`: @NotBlank String,
        inputStream: @NotNull InputStream,
        contentType: @NotBlank String
    ): ObjectWriteResponse {
        return minioClient.putObject(
            PutObjectArgs
                .builder()
                .bucket(bucket)
                .`object`(`object`)
                .stream(inputStream, inputStream.available().toLong(), -1)
                .contentType(contentType)
                .build()
        )
    }

    fun listBucketNames(): List<String> {
        return listBuckets().stream().map { obj: Bucket -> obj.name() }.toList()
    }

    fun listBuckets(): List<Bucket> {
        return minioClient.listBuckets()
    }

    fun removeBucket(bucket: @NotBlank String?): Boolean {
        if (!bucketExists(bucket)) {
            return false
        }
        val myObjects = listObjects(bucket)
        if (CollUtil.isEmpty(myObjects)) {
            return false
        }
        for (result in myObjects!!) {
            val item = result.get()
            // If item has files, fail to remove
            if (item.size() > 0) {
                return false
            }
        }
        // If bucket is empty, then it can be removed
        minioClient.removeBucket(RemoveBucketArgs.builder().bucket(bucket).build())
        return !bucketExists(bucket)
    }

    fun listObjectNames(bucket: @NotBlank String): List<String> {
        val objectNames = Lists.newArrayList<String>()
        if (!bucketExists(bucket)) {
            return objectNames
        }
        val objects = listObjects(bucket)
        if (CollUtil.isEmpty(objects)) {
            return objectNames
        }
        for (result in objects!!) {
            val item = result.get()
            objectNames.add(item.objectName())
        }
        return objectNames
    }

    fun listObjects(bucket: @NotBlank String?): Iterable<Result<Item>>? {
        return if (!bucketExists(bucket)) {
            null
        } else minioClient.listObjects(ListObjectsArgs.builder().bucket(bucket).build())
    }

    fun getObject(bucket: @NotBlank String, `object`: @NotBlank String): GetObjectResponse? {
        if (!bucketExists(bucket)) {
            return null
        }
        val statObjectResponse = statObject(bucket, `object`)
        return if (statObjectResponse == null || statObjectResponse.size() == 0L) {
            null
        } else minioClient.getObject(GetObjectArgs.builder().bucket(bucket).`object`(`object`).build())
    }

    fun getObject(
        bucket: @NotBlank String,
        `object`: @NotBlank String,
        offset: @Min(0L) Long,
        length: @NotNull @Min(0L) Long
    ): GetObjectResponse? {
        if (!bucketExists(bucket)) {
            return null
        }
        val statObjectResponse = statObject(bucket, `object`)
        return if (statObjectResponse == null || statObjectResponse.size() == 0L) {
            null
        } else minioClient.getObject(
            GetObjectArgs.builder().bucket(bucket).`object`(`object`).offset(offset).length(length).build()
        )
    }

    fun removeObjects(bucket: @NotBlank String?, `object`: @NotBlank String?): Boolean {
        if (!bucketExists(bucket)) {
            return false
        }
        minioClient.removeObject(RemoveObjectArgs.builder().bucket(bucket).`object`(`object`).build())
        return true
    }

    fun removeObjects(bucket: @NotBlank String, objects: @NotEmpty List<String>): List<String> {
        val deleteErrorNameList = mutableListOf<String>()
        if (bucketExists(bucket)) {
            val deleteObjectList = objects.stream().map { name: String? -> DeleteObject(name) }
                .toList()
            val results = minioClient.removeObjects(
                RemoveObjectsArgs.builder().bucket(bucket).objects(deleteObjectList).build()
            )
            for (result in results) {
                val error = result.get()
                deleteErrorNameList.add(error.objectName())
            }
        }
        return deleteErrorNameList
    }

    fun getPresignedObjectUrl(
        bucket: @NotBlank String,
        `object`: @NotBlank String,
        method: @NotNull Method,
        expiration: @NotNull @Min(1L) @Max(DEFAULT_EXPIRY_TIME.toLong()) Int = DEFAULT_EXPIRY_TIME
    ): String? {
        return if (!bucketExists(bucket)) {
            null
        } else minioClient.getPresignedObjectUrl(
            GetPresignedObjectUrlArgs
                .builder()
                .bucket(bucket)
                .`object`(`object`)
                .method(method)
                .expiry(expiration)
                .build()
        )
    }

    fun statObject(bucket: @NotBlank String, `object`: @NotBlank String): StatObjectResponse? {
        return if (!bucketExists(bucket)) {
            null
        } else minioClient.statObject(StatObjectArgs.builder().bucket(bucket).`object`(`object`).build())
    }

    fun getPresignedObjectUrl(bucket: @NotBlank String, `object`: @NotBlank String): String? {
        return if (!bucketExists(bucket)) {
            null
        } else minioClient.getPresignedObjectUrl(
            GetPresignedObjectUrlArgs.builder().bucket(bucket).`object`(`object`).build()
        )
    }

    fun composeObject(
        bucket: @NotBlank String,
        `object`: @NotBlank String,
        sources: @NotEmpty MutableList<ComposeSource>,
        headers: Map<String, String>?
    ): ObjectWriteResponse? {
        return if (!bucketExists(bucket)) {
            null
        } else minioClient.composeObject(
            ComposeObjectArgs.builder().bucket(bucket).`object`(`object`).sources(sources).headers(headers).build()
        )
    }
}
